Write the 1.7 release notes

Gallaecio opened the pull request:
Current coverage: https://github.com/scrapy/scrapy/commit/44eb21aa51cc0212e68dc0709aaa5c10bfe64e7e

Includes a note about dropping Python 2 support in Scrapy 2.0.

Fixes #3872.

kmike commented:
hey @Gallaecio! I also tried to put all pull requests for new changes in thee 1.7 milestone a few days ago (https://github.com/scrapy/scrapy/milestone/22), maybe this would help you. Not 100% sure all PRs are there though, I could have missed something, or added something which shouldn't be there.

Gallaecio commented on docs/news.rst:
```diff
@@ -3,6 +3,101 @@
+    :issue:`3635`)
+
+*   A new :setting:`FEED_STORAGE_S3_ACL` setting allows defining a custom ACL
+    for feeds exported to Amazon S3 by your project (:issue:`3607`)
```
```suggestion
    for feeds exported to Amazon S3 (:issue:`3607`)
```

Gallaecio commented on docs/news.rst:
```diff
@@ -3,6 +3,101 @@
+*   :class:`~scrapy.crawler.Crawler`,
+    :class:`CrawlerRunner.crawl <scrapy.crawler.CrawlerRunner.crawl>` and
+    :class:`CrawlerRunner.create_crawler <scrapy.crawler.CrawlerRunner.create_crawler>`
+    now fail gracefully if they receive a :class:`scrapy.Spider` object
```
Add “instead of a Spider class” or similar.

Gallaecio commented:
> hey @Gallaecio! I also tried to put all pull requests for new changes in thee 1.7 milestone a few days ago (https://github.com/scrapy/scrapy/milestone/22), maybe this would help you. Not 100% sure all PRs are there though, I could have missed something, or added something which shouldn't be there.

Thanks! I’ll complete the release notes based on commits just to make sure I don’t miss anything, and then compare the result to the milestone to ensure that everything fits.

Gallaecio commented:
https://github.com/readthedocs/sphinx_rtd_theme/issues/705 is making some list items look rather badly, I hope they fix it soon.

Gallaecio commented:
Current coverage:
- Commits until https://github.com/scrapy/scrapy/commit/3adf09bc4f1468f54a703481d6d7ebfe25b6f232
- Pending pull requests https://github.com/scrapy/scrapy/pull/3794 and https://github.com/scrapy/scrapy/pull/3578

While I’m confident about the coverage (as in I don’t think any issue is missing), I’m more than grateful for any feedback regarding everything else (e.g. content organization, header naming).

kmike commented on docs/news.rst:
```diff
@@ -3,6 +3,269 @@
+    (:issue:`1138`, :issue:`3563`)
+
+*   A new :class:`~scrapy.http.JSONRequest` class offers a more convenient way
+    to build JSON requests (:issue:`3504`, :issue:`3505`)
```
What do you think about moving most prominent features (visible to users) to the top? Maybe new queue, cb_kwargs and JSONRequest?

Gallaecio commented:
Moved those 3 new features to the top, simplified the list of deprecations further to remove as much visual clutter as possible without loss of information, and covered #3692.

Current coverage: https://github.com/scrapy/scrapy/commit/4c755f1ddc6593b8428e8355ab00e4e9691aac63 + https://github.com/scrapy/scrapy/pull/3578

Gallaecio commented:
Added a note about Scrapy 2.0 dropping support for Python 2.

Gallaecio commented:
I’ll keep the description up-to-date with the release notes coverage.

elacuesta commented:
Hi there @Gallaecio, @kmike, have we reached feature freeze for 1.7 or is there a chance to include https://github.com/scrapy/scrapy/pull/3712 and https://github.com/scrapy/scrapy/pull/2400? Thanks!

elacuesta commented on docs/news.rst:
```diff
@@ -2471,12 +2777,22 @@ First release of Scrapy.
-.. _resource: https://docs.python.org/2/library/resource.html
-.. _queuelib: https://github.com/scrapy/queuelib
+.. _Creating a pull request: https://help.github.com/en/articles/creating-a-pull-request
 .. _cssselect: https://github.com/SimonSapin/cssselect
```
This is currently redirecting to `scrapy/cssselect`, but I think it would be good to update it either way.

kmike commented on docs/news.rst:
```diff
@@ -3,6 +3,312 @@
+
+*   ``429`` is now part of the :setting:`RETRY_HTTP_CODES` setting by default.
+
+    This change is **backward incompatible**. If you don’t want to retry
```
I wonder if we should have a separate section for this, at the top: backwards incompatible changes + a link to a section with deprecation removals, so that users have a single place to read before the upgrade.

Gallaecio commented:
Both suggestions applied.

I’ve also made some changes to the documentation from #3829 which include fixes for link syntax issues that I overlooked.

kmike commented on docs/news.rst:
```diff
@@ -3,6 +3,322 @@
+~~~~~~~~~~~~~
+
+*   ``scrapy.utils.gz.is_gzipped`` is now **deprecated**. Use
+    ``scrapy.utils.gz.gzip_magic_number`` instead.
```
should it be moved to Deprecations section?

Gallaecio replied:
Right! :man_facepalming:

Gallaecio replied:
Done.

I’ve also moved the two deprecation comments from _New features_ to _Deprecations_ as well.

kmike commented on docs/news.rst:
```diff
@@ -3,6 +3,319 @@
+
+*   The ``queuelib.PriorityQueue`` value for the
+    :setting:`SCHEDULER_PRIORITY_QUEUE` setting is deprecated. Use
+    :class:`scrapy.pqueues.ScrapyPriorityQueue` instead.
```
This is actually backwards incompatible: only exact queuelib.PriorityQueue class still works, its custom subclasses don't. See https://github.com/scrapy/scrapy/blob/df68c4b9b1789cc2c7f800f8e4b9f6e5aca7b7af/scrapy/core/scheduler.py#L59

The main API change is that priority queue class now needs to support Request objects, not just general Python data structures; compare https://github.com/scrapy/scrapy/blob/df68c4b9b1789cc2c7f800f8e4b9f6e5aca7b7af/scrapy/pqueues.py#L93 and https://github.com/scrapy/queuelib/blob/master/queuelib/pqueue.py.

I'm not sure how detailed should be our deprecation docs on this, as this API was not documented before, if I'm not mistaken.

Gallaecio replied:
I’ve added a short entry to the backward incompatibility section, please have a look and share your thoughts.

elacuesta commented:
> Added a note about Scrapy 2.0 dropping support for Python 2.

Can't wait for this. Don't get me wrong, I think `six` does an awesome job, but it'll be good to clean up the codebase.

Gallaecio commented:
Updated to fix #3872.

Gallaecio commented:
As part of the last updated, I’ve:
- Installed the InterSphinx extension to easily link to external documentations also built using Sphinx, like Python’s. ``ChainMap`` served as an excuse for it :slightly_smiling_face: 
- Grouped together a few of the documentation entries, to make the documentation section shorter.
- Moved the entry about generating documentation coverage data to *Other changes*, as it’s meant for contributors, not users.
